File: src/Tirax.KMS/ChangeLogs.fs
namespace Tirax.KMS

open System.Runtime.CompilerServices
open RZ.FSharp.Extension
open Domain

[<Struct; IsReadOnly>]
type ModelOperationType<'T> =
    | Add of add:'T
    | Update of update_old:'T * update_new:'T
    | Delete of delete:'T

[<Extension>]
type ModelOperationTypeExtensions =
    [<Extension>]
    static member inline apply<'T when 'T: comparison> (operation :ModelOperationType<'T>, s :Set<'T>) =
        match operation with
        | Add x         -> s.Add(x)
        | Update (_, x) -> if not (s.Contains x) then s.Add(x) else s
        | Delete x      -> s.Remove(x)
        
    [<Extension>]
    static member inline apply(operation :ModelOperationType<'T>, m :Map<'K,'T>, keyIdentifier :'T -> 'K) =
        match operation with
        | Add x         -> m.Add(keyIdentifier x, x)
        | Update (_, x) -> m.Change(keyIdentifier x, constant (Some x))
        | Delete x      -> m.Remove(keyIdentifier x)
        
    [<Extension>]
    static member inline applyKeyValue(operation :ModelOperationType<struct ('K*'T)>, m :Map<'K,'T>) =
        match operation with
        | Add struct (k,v)         -> m.Add(k, v)
        | Update (_, struct (k,v)) -> m.Change(k, constant (Some v))
        | Delete struct (k,_)      -> m.Remove(k)
    
[<Struct; IsReadOnly>]
type ModelChange =
    | Tag           of tag:ModelOperationType<ConceptTag>
    | ConceptChange of concept:ModelOperationType<Concept>
    | OwnerChange   of owner:ModelOperationType<struct (ConceptId * ConceptId list)>
    
type ChangeLogs = ModelChange seq

File: src/Tirax.KMS.Core/Server/ChangeLogs.cs
using Tirax.KMS.Domain;
// ReSharper disable InconsistentNaming

namespace Tirax.KMS.Server;

public abstract record ModelOperationType<T>
{
    public sealed record Add(T Value) : ModelOperationType<T>;
    public sealed record Update(T Old, T New) : ModelOperationType<T>;
    public sealed record Delete(T Value) : ModelOperationType<T>;
}

public static class ModelOperationType
{
    public static ModelOperationType<T> Add<T>(T value) => new ModelOperationType<T>.Add(value);
    public static ModelOperationType<T> Update<T>(T old, T @new) => new ModelOperationType<T>.Update(old, @new);
    public static ModelOperationType<T> Delete<T>(T value) => new ModelOperationType<T>.Delete(value);
}

public static class ModelOperationTypeExtensions
{
    public static LanguageExt.HashSet<T> Apply<T>(this ModelOperationType<T> operation, LanguageExt.HashSet<T> s) =>
        operation switch{
            ModelOperationType<T>.Add x    => s.Add(x.Value),
            ModelOperationType<T>.Update x => s.Contains(x.New) ? s : s.Add(x.New),
            ModelOperationType<T>.Delete x => s.Remove(x.Value),
            _                              => throw new NotSupportedException()
        };

    public static Map<K, T> Apply<K, T>(this ModelOperationType<T> operation, Map<K, T> m, Func<T, K> keyIdentifier) =>
        operation switch{
            ModelOperationType<T>.Add x    => m.Add(keyIdentifier(x.Value), x.Value),
            ModelOperationType<T>.Update x => m.AddOrUpdate(keyIdentifier(x.New), _ => x.New, x.New),
            ModelOperationType<T>.Delete x => m.Remove(keyIdentifier(x.Value)),
            _                              => throw new NotSupportedException()
        };

    public static Map<K, T> Apply<K, T>(this ModelOperationType<(K,T)> operation, Map<K, T> m) =>
        operation switch{
            ModelOperationType<(K,T)>.Add x    => x.Value.Map(m.Add),
            ModelOperationType<(K,T)>.Update x => m.AddOrUpdate(x.New.Item1, _ => x.New.Item2, x.New.Item2),
            ModelOperationType<(K,T)>.Delete x => m.Remove(x.Value.Item1),
            _                                  => throw new NotSupportedException()
        };

}

public abstract record ModelChange
{
    public sealed record Tag(ModelOperationType<ConceptTag> Value) : ModelChange;
    public sealed record ConceptChange(ModelOperationType<Concept> Concept) : ModelChange;
    public sealed record OwnerChange(ModelOperationType<(ConceptId ConceptId, Seq<ConceptId> Owners)> Owner) : ModelChange;
    public sealed record LinkObjectChange(ModelOperationType<LinkObject> Link) : ModelChange;
    
    public static ModelChange NewTag(ModelOperationType<ConceptTag> Value) => new Tag(Value);
    public static ModelChange NewConceptChange(ModelOperationType<Concept> Concept) => new ConceptChange(Concept);
    public static ModelChange NewOwnerChange(ModelOperationType<(ConceptId ConceptId, Seq<ConceptId> Owners)> Owner) => new OwnerChange(Owner);
    public static ModelChange NewLinkObjectChange(ModelOperationType<LinkObject> Link) => new LinkObjectChange(Link);
}

public readonly record struct ChangeLogs(Seq<ModelChange> Value);